posix: eventfd: Fix unsetting internal flags in ioctl

celinakalus opened the pull request:
Commit e6eb0a705bc8 ("posix: eventfd: revise locking, signaling, and allocation") introduced a regression where the internal flags of an event file descriptor would be erased when calling the F_SETFL ioctl operation.

This includes the flag EFD_IN_USE_INTERNAL which determines whether this file descriptor has been opened, thus effectively closing the eventfd whenever one tries to change a flag.

Fixes #71399

celinakalus committed:
posix: eventfd: Fix unsetting internal flags in ioctl

Commit e6eb0a705bc8 ("posix: eventfd: revise locking, signaling, and
allocation") introduced a regression where the internal flags of an
event file descriptor would be erased when calling the F_SETFL ioctl
operation.

This includes the flag EFD_IN_USE_INTERNAL which determines whether
this file descriptor has been opened, thus effectively closing the
eventfd whenever one tries to change a flag.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

jukkar commented on lib/posix/options/eventfd.c:
```diff
@@ -247,7 +247,9 @@ static int eventfd_ioctl_op(void *obj, unsigned int request, va_list args)
 			ret = -1;
 		} else {
-			efd->flags = flags;
+			int prev_flags = efd->flags & ~EFD_FLAGS_SET_INTERNAL;
+
+			efd->flags = flags | prev_flags;
```
We could avoid extra variable by this:
```
efd->flags = (efd->flags & ~EFD_FLAGS_SET_INTERNAL ) | flags;
```
The line is still readable and not even that long.

celinakalus replied:
Thanks for your feedback. Personally I prefer the separate variable, but if that is an issue, of course I'll change it.

cfriedt reviewed (commented):
The changes look fine (I'm not terribly opinionated about the additional variable).

However, since it has been identified as a regression, it would be much appreciated if you could add a second commit to this PR to test for the correct / expected functionality in the eventfd testsuite.

With that, at least CI failures will prevent this type of regression in the future.

Thanks for the fix 😊

ycsin commented:
..and probably a bug report so that it can be backported?

celinakalus committed:
tests: posix: eventfd: Add ioctl F_SETFL test

Add a test to protect against future regressions in the ioctl F_SETFL
operation of eventfd. Flags are set and unset and validity of the file
descriptor is checked by reading and writing.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus commented:
I added a test and a bug report so it can be backported.

cfriedt reviewed (approved):
LGTM 👍
